mpc_eval.py

```diff
@@ -124,51 +121,50 @@ def diverse_sampling(code):
-                    target_now_codes = image_encoder(target_now).detach()
-                    now_codes = torch.cat([state_now_codes, target_now_codes], dim=1).squeeze()
-
-                    diverse_now_codes, now_noises = diverse_sampling(now_codes)
-                    diverse_now_codes, now_noises = diverse_now_codes[..., None, None], now_noises[..., None, None]
-
-                    action_now_hat = generator(diverse_now_codes.view(-1, diverse_now_codes.size(2)))
-                    action_now_hat = action_now_hat.view(batch_size,-1,4)
-                    
-                    if ts==0:
-                        action_now_taken = action_now_hat
-                    #forward model
-                    code_fwd, feats_fwd = fwd_model_encoder(state_cur_fwd)
-                    state_action_concate = torch.cat([code_fwd, action_now_hat.squeeze(1)], dim=1)
-                    state_action_concate = state_action_concate.unsqueeze(2).unsqueeze(3)
-                    state_fut_hat = fwd_model_decoder(state_action_concate, feats_fwd)
-                    state_cur_fwd = state_fut_hat
-                    
-                err_now = mse(state_fut_hat,state_target)
+                diverse_now_codes, now_noises = diverse_sampling(now_codes)
+                diverse_now_codes, now_noises = diverse_now_codes[..., None, None], now_noises[..., None, None]
+                action_now_hat = generator(diverse_now_codes.view(-1, diverse_now_codes.size(2)))
+                action_now_hat = action_now_hat.view(rollouts,-1,4)
+                if ts==0:
+                    action_now_taken = action_now_hat
+
+                #forward model
+                state_fut_hat = fwd_model_autoencoder(state_cur_fwd,action_now_hat.squeeze(1))
+                state_cur_fwd = state_fut_hat
+            
+            best_act_ind = 0
+            for ro in range(rollouts):
+                err_now = mse(state_fut_hat[ro],state_target[0])
                 if err_now<min_error:
                     min_error = err_now
-                best_action_so_far = action_now_taken
+                    best_act_ind = ro
+            best_action_so_far = action_now_taken[best_act_ind]
+            
+            state_cur_mpc = fwd_model_autoencoder(state_cur_mpc,best_action_so_far.squeeze(1))
             
-            code_fwd, feats_fwd = fwd_model_encoder(state_cur_mpc)
-            state_action_concate = torch.cat([code_fwd, best_action_so_far.squeeze(1)], dim=1)
-            state_action_concate = state_action_concate.unsqueeze(2).unsqueeze(3)
-            state_cur_mpc = fwd_model_decoder(state_action_concate, feats_fwd)
             best_action_list.append(best_action_so_far)
 
             image_error = mse(state_cur_mpc, state_fut)
 
             # Cumulative action error with diverse samples
             image_error_sum += image_error
             step += 1
-        action_hat = torch.cat(best_action_list,dim=1)
+        action_hat = torch.cat(best_action_list,dim=0)
```

Comment:
Why did this change?

Reply:
It isn't used as such, was just for MSE error calculation. Had a shape error so I had to flatten it instead of stack.